### PR TITLE
[LLVMCPU] Enable masking vectorization for tensor.pad ops.

### LINKED ISSUE #13487: CPU padding approach results in huge stack allocation
https://github.com/openxla/iree/pull/13420 fixes the cases that all dims are unaligned, but it does not cover all the cases. It still can result in huge stack allocation if outer pad ops are created but not inner pad ops. In this case, we won't be able to fold the pad op into the innermost loop, which results in huge stack allocations.

To repro: `iree-opt --split-input-file --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target)))" ~/pipeline.mlir`

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {
  cpu_features = "",
  data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
  native_vector_size = 64 : index,
  target_triple = "x86_64-unknown-unknown-eabi-elf"}>
#pipeline_layout5 = #hal.pipeline.layout<push_constants = 2, sets = [
  #hal.descriptor_set.layout<0, bindings = [
    #hal.descriptor_set.binding<0, storage_buffer>,
    #hal.descriptor_set.binding<1, storage_buffer>,
    #hal.descriptor_set.binding<2, storage_buffer>]
  >]>
hal.executable private @matmul {
  hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
    hal.executable.export public @matmul ordinal(0) layout(#pipeline_layout5) {
    ^bb0(%arg0: !hal.device):
      %x, %y, %z = flow.dispatch.workgroup_count_from_slice
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @matmul() {
        %cst = arith.constant 0.000000e+00 : f32
        %0 = hal.interface.constant.load[0] : i32
        %1 = hal.interface.constant.load[1] : i32
        %2 = hal.interface.constant.load[2] : i32
        %3 = hal.interface.constant.load[3] : i32
        %4 = arith.index_castui %0 {stream.alignment = 128 : index, stream.values = [0 : index, 131712 : index]} : i32 to index
        %5 = arith.index_castui %1 {stream.alignment = 64 : index, stream.values = [576704 : index, 1763072 : index]} : i32 to index
        %6 = arith.index_castui %2 {stream.alignment = 64 : index, stream.values = [908480 : index, 2094848 : index]} : i32 to index
        %7 = arith.index_castui %3 {stream.alignment = 128 : index, stream.values = [2304 : index, 134016 : index]} : i32 to index
        %8 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%4) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<1x576xf32>>
        %9 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%5) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<576x144xf32>>
        %10 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%6) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<1x144xf32>>
        %11 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%7) : !flow.dispatch.tensor<writeonly:tensor<1x144xf32>>
        %12 = flow.dispatch.tensor.load %8, offsets = [0, 0], sizes = [1, 576], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<1x576xf32>> -> tensor<1x576xf32>
        %13 = flow.dispatch.tensor.load %9, offsets = [0, 0], sizes = [576, 144], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<576x144xf32>> -> tensor<576x144xf32>
        %14 = flow.dispatch.tensor.load %10, offsets = [0, 0], sizes = [1, 144], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<1x144xf32>> -> tensor<1x144xf32>
        %15 = tensor.empty() : tensor<1x144xf32>
        %16 = linalg.fill ins(%cst : f32) outs(%15 : tensor<1x144xf32>) -> tensor<1x144xf32>
        %17 = linalg.matmul ins(%12, %13 : tensor<1x576xf32>, tensor<576x144xf32>) outs(%16 : tensor<1x144xf32>) -> tensor<1x144xf32>
        %18 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%17, %14 : tensor<1x144xf32>, tensor<1x144xf32>) outs(%15 : tensor<1x144xf32>) {
        ^bb0(%in: f32, %in_0: f32, %out: f32):
          %19 = arith.addf %in, %in_0 : f32
          %20 = arith.maxf %19, %cst : f32
          linalg.yield %20 : f32
        } -> tensor<1x144xf32>
        flow.dispatch.tensor.store %18, %11, offsets = [0, 0], sizes = [1, 144], strides = [1, 1] : tensor<1x144xf32> -> !flow.dispatch.tensor<writeonly:tensor<1x144xf32>>
        return
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by hanhanW**
This also fixes an issue in DoubleTilingPadExpert when enabling masking. The approach is expected to generate tensor.pad op at innermost region, which kicks in canonicalization patterns to fold outer tensor.pad ops away. If they are not folded away, the vectorizer will create a mask for the op and they result in huge stack allocation. It forces LLVMCPUTensorPad pass to preserve the tensor.pad op at innermost region, so the canonicalization pattern can remove the outer tensor.pad ops as expected.

Fixes https://github.com/openxla/iree/issues/13487

### CODE REVIEW

**Comment by hanhanW** [comment]
It depends on https://reviews.llvm.org/D150227

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTensorPad.cpp [hunk 1]
```diff
@@ -36,22 +36,33 @@ class LLVMCPUTensorPadPass : public LLVMCPUTensorPadBase<LLVMCPUTensorPadPass> {
...
+      nofold = true;
       break;
   };
   SmallVector<linalg::LinalgOp> candidates;
   funcOp.walk([&](linalg::LinalgOp op) { candidates.push_back(op); });
   for (auto linalgOp : candidates) {
     IRRewriter rewriter(context);
     LLVM_DEBUG(llvm::dbgs() << "candidate: " << linalgOp);
+
+    if (option == LLVMCPUTensorPadOption::ParallelDims &&
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTensorPad.cpp:62]: Is this related to vectorization of pad operations?

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTensorPad.cpp:62]: half half. It does not directly relate to masking vectorization. Masking vectorization exposes an issue in padding approach; this is the fix. It was not an issue because we apply a sequence of patterns to vectorize pad op, and they got folded away somehow. It is a new issue because we vectorize pad op in one-step using masking. It's easier to me if we have them in one PR. I can break it into two PRs if you have concerns.

**Review by MaheshRavishankar — APPROVED** [review]
Looks fine to me.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTensorPad.cpp (hunk 1 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTensorPad.cpp:62]: Could you please add a comment about what this is for?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorization.cpp [hunk 2]
```diff
@@ -234,16 +234,33 @@ void LLVMCPUVectorizationPass::runOnOperation() {
...
+      auto ty = padOp.getResultType();
+      // TODO(hanchung): Infer the vector sizes for pad op after
+      // maskedVectorize method allows dynamic result shapes.
+      if (!ty.hasStaticShape()) continue;
+      SmallVector<int64_t> vectorSizes(ty.getShape().begin(),
+                                       ty.getShape().end());
+      FailureOr<vector::TransferWriteOp> maybeWriteOp =
+          linalg::maskedVectorize(rewriter, padOp, vectorSizes);
+      if (failed(maybeWriteOp)) {
+        continue;
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorization.cpp:261]: Unrelated to this patch but we should unify these interfaces upstream.

**Review by dcaballe — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
I am assuming that this will land after the LLVM commit comes through

**Comment by hanhanW** [comment]
I think this should get in after we have proper tensor.pad masking vectorization. The current version is creating a mask, it does not reuse existing mask when possible. Mark it to draft until we have proper vectorization.

**Comment by dcaballe** [comment]
Masks won't be reused across linalg/tensor ops as the vectorization state is not preserved across ops. Redundant masks are optimized away so that shouldn't be a problem.

**Comment by hanhanW** [comment]
> Masks won't be reused across linalg/tensor ops as the vectorization state is not preserved across ops. Redundant masks are optimized away so that shouldn't be a problem.

Good to know, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorization.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorization.cpp:261]: follow-up in https://reviews.llvm.org/D150495
